### PR TITLE
Add SelfRemove proposal to MLS extensions

### PR DESCRIPTION
**Pull request by rohan-wire**
As discussed during IETF 117, this PR corresponds to the text from draft-mahy-mls-selfremove.

### CODE REVIEW

File: draft-ietf-mls-extensions.md [hunk 1]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
 `application_content` is interpreted as the first MediaType specified in
 `required_media_types`.
 
+## SelfRemove Proposal
+
+The design of the MLS protocol prevents a member of
+an MLS group from removing itself immediately from the group. To cause
+an immediate change in the group, a member must send a Commit message.
```
Comment (kkohbrok) [inline draft-ietf-mls-extensions.md:498]: The first of these three sentences seems to be the conclusion of the second and third. Maybe change up the order a bit?

File: draft-ietf-mls-extensions.md [hunk 2]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
...
+However the sender of a Commit message knows the keying material of the
+new epoch and therefore needs to be part of the group. Instead a member
+wishing to remove itself can send a Remove Proposal and wait for another
+member to Commit its Proposal.
+
+Unfortunately, MLS clients that join via an External Commit ignore
+pending, but otherwise valid, Remove Proposals. The member trying to remove itself has
+to monitor the group and send a new Remove Proposal in any new epoch until the member is
```
Comment (kkohbrok) [inline draft-ietf-mls-extensions.md:506]:
```suggestion
to monitor the group and send a new Remove Proposal in every new epoch until the member is
```

File: draft-ietf-mls-extensions.md [hunk 3]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
...
+new participants are trying to join (often on the hour).
+
+This section describes a new `SelfRemove` Proposal extension type. It is
+designed to be included in External Commits.
+
+### Extension Description
+
+This document specifies a new MLS Proposal type called `SelfRemove`. Its syntax
+is described using the TLS Presentation Language [@!RFC8446] below (its contents
+is an empty struct). It is allowed in External Commits and requires an UpdatePath.
```
Comment (kkohbrok) [inline draft-ietf-mls-extensions.md:521]:
```suggestion
are an empty struct). It is allowed in External Commits and requires an UpdatePath.
```
This should be plural, shouldn't it?

File: draft-ietf-mls-extensions.md [hunk 4]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
...
+        case self_remove:              SelfRemove;
+    };
+} Proposal;
+~~~
+
+The description of behavior below only applies if all the
+members of a group support this extension in their
+capabilities; such a group is a "self-remove-capable group".
+
+An MLS client which implements this specification can send a
```
Comment (kkohbrok) [inline draft-ietf-mls-extensions.md:547]:
```suggestion
An MLS client which supports this extension can send a
```

File: draft-ietf-mls-extensions.md [hunk 5]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
...
+SelfRemove Proposal whenever it would like to remove itself
+from a self-remove-capable group. Because the point of a
+SelfRemove Proposal is to be available to external joiners
+(which are not yet members), these proposals MUST be sent
+as an MLS PublicMessage.
```
Comment (kkohbrok) [inline draft-ietf-mls-extensions.md:552]:
```suggestion
as MLS PublicMessages.
```

File: draft-ietf-mls-extensions.md [hunk 6]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
...
+
+Whenever a member receives a SelfRemove Proposal, it includes
+it along with any other pending Propsals when sending a Commit.
+It already MUST send a Commit of pending Proposals before sending
+new application messages.
+
+When a member receives a Commit with an embedded SelfRemove Proposal,
```
Comment (kkohbrok) [inline draft-ietf-mls-extensions.md:559]: "embedded" (at least to me) implies that it's by-value. Maybe
```suggestion
When a member receives a Commit referencing one or more SelfRemove Proposals,
```

File: draft-ietf-mls-extensions.md [hunk 7]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
...
+it treats the proposal like a Remove Proposal, except the leaf node to remove
+is determined by looking in the Sender `leaf_index` of the original Proposal.
+The member is able to verify that the Sender was a member.
+
+Whenever a new joiner is about to join a self-remove-capable group with an
+External Commit, the new joiner MUST fetch any pending SelfRemove Proposals
+along with the GroupInfo object, and include the SelfRemove Proposals
+in its External Commit by reference. (An ExternalCommit can contain zero or
+more SelfRemove proposals). The new joiner validates the SelfRemove
```
Comment (kkohbrok) [inline draft-ietf-mls-extensions.md:568]:
```suggestion
more SelfRemove proposals). The new joiner MUST validate the SelfRemove
```

File: draft-ietf-mls-extensions.md [hunk 8]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
...
+Proposal before including it by reference, except that it skips the validation
+of the `membership_tag` because a non-member cannot verify membership.
+
+During validation, SelfRemove proposals are processed after Update proposals
+and before Remove proposals. If there is a pending SelfRemove proposal for a specific
+leaf node and a pending Remove proposal for the same leaf node, the Remove proposal is
+invalid. A client MUST NOT issue more than one SelfRemove proposal per epoch.
+
+The MLS Distribution Service (DS) needs to validate SelfRemove Proposals it
```
Comment (kkohbrok) [inline draft-ietf-mls-extensions.md:577]:
```suggestion
The MLS Delivery Service (DS) needs to validate SelfRemove Proposals it
```

Comment (kkohbrok) [reply draft-ietf-mls-extensions.md:577]: "needs to" maybe this should be "SHOULD"?

File: draft-ietf-mls-extensions.md [hunk 9]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
...
+receives (except that it cannot validate the `membership_tag`). If the DS
+provides a GroupInfo object to an external joiner, the DS SHOULD attach any
+SelfRemove proposals known to the DS to the GroupInfo object.
+
+As with Remove proposals, clients need to be prepared to receive the Commit
```
Comment (kkohbrok) [inline draft-ietf-mls-extensions.md:582]: This is a bit unclear to me. What does it mean for a client to be prepared?

**Review by kkohbrok — CHANGES_REQUESTED** [review]
A very useful extension and the PR looks good in general.

I left a few comments inline. Also, the text should probably be wrapped to 80(?) characters.

File: draft-ietf-mls-extensions.md (hunk 9 above)
Comment (rohan-wire) [inline draft-ietf-mls-extensions.md:582]:
```suggestion
As with Remove proposals, clients need to be able to receive a Commit
```

File: draft-ietf-mls-extensions.md [hunk 10]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
...
+is a member trying to remove itself from a conference call as several dozen
+new participants are trying to join (often on the hour).
+
+This section describes a new `SelfRemove` Proposal extension type. It is
+designed to be included in External Commits.
+
+### Extension Description
+
+This document specifies a new MLS Proposal type called `SelfRemove`. Its syntax
+is described using the TLS Presentation Language [@!RFC8446] below (its contents
```
Comment (rohan-wire) [inline draft-ietf-mls-extensions.md:520]:
```suggestion
is described using the TLS Presentation Language [@!RFC8446] below (its content
```

File: draft-ietf-mls-extensions.md [hunk 11]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
 `application_content` is interpreted as the first MediaType specified in
 `required_media_types`.
 
+## SelfRemove Proposal
+
+The design of the MLS protocol prevents a member of
+an MLS group from removing itself immediately from the group. To cause
```
Comment (rohan-wire) [inline draft-ietf-mls-extensions.md:497]:
```suggestion
an MLS group from removing itself immediately from the group. (To cause
```

File: draft-ietf-mls-extensions.md [hunk 12]
```diff
@@ -483,6 +491,99 @@ The `media_type` MAY be zero length, in which case, the media type of the
...
+an immediate change in the group, a member must send a Commit message.
+However the sender of a Commit message knows the keying material of the
+new epoch and therefore needs to be part of the group. Instead a member
```
Comment (rohan-wire) [inline draft-ietf-mls-extensions.md:500]:
```suggestion
new epoch and therefore needs to be part of the group.) Instead a member
```

**Review by kkohbrok — APPROVED** [review]

**Pull request merged**